dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java

```diff
@@ -610,13 +610,33 @@ private Workflow createOrGetWorkflow(Class workflowType, String repository, User
             }
         }
 
+        // Check that the subclass is the same as the entry to update
+        // For services, we must compare the descriptor type subclasses
+        // For workflows and tools, we must compare the descriptor types (languages)
+        // The `convertShortName` methods throw `UnsupportedOperationException` when passed an unknown subclass
+        try {
+            if (workflowType == Service.class) {
+                checkSame(workflowToUpdate.getDescriptorTypeSubclass(), DescriptorLanguageSubclass.convertShortNameStringToEnum(subclass), "descriptor type subclass", "service");
+            } else {
+                checkSame(workflowToUpdate.getDescriptorType(), DescriptorLanguage.convertShortStringToEnum(subclass), "descriptor language (subclass)", workflowToUpdate instanceof AppTool ? "tool" : "workflow");
+            }
+        } catch (UnsupportedOperationException e) {
+            throw new CustomWebApplicationException(String.format("Unknown subclass '%s'", subclass), HttpStatus.SC_BAD_REQUEST);
+        }
+
         if (user != null) {
             workflowToUpdate.getUsers().add(user);
         }
 
         return  workflowToUpdate;
     }
 
+    private <T> void checkSame(T currentValue, T newValue, String valueDescription, String entryDescription) {
+        if (!Objects.equals(currentValue, newValue)) {
+            throw new CustomWebApplicationException(String.format("The %s of the original %s ('%s') and all of its versions must be the same.", valueDescription, entryDescription, currentValue), HttpStatus.SC_BAD_REQUEST);
```

Comment:
I think the message could be more specific; instead of just saying they must be the same, say what they are. Something like, very crudely phrased and should be improved if you choose to go down this route, "You can't add a CWL version to a WDL workflow" is more actionable to an end user than "The descriptor language (subclass) of the original workflow (CWL) and all of its versions must be the same".

Reply:
I changed the message to be a pleasing mashup of your suggestion and the more technical explanation.